### PR TITLE
project query enhance

### PR DESCRIPTION
**Pull request by k9845**
Addresses https://github.com/the-deep/server/issues/893

## Changes
Added
  - membersCount
  - sourcesCount
  - entriesCount
  - userStatus

## This PR doesn't introduce any:

- [x] temporary files, auto-generated files or secret keys
- [x] n+1 queries
- [x] flake8 issues
- [x] `print`
- [x] typos
- [x] unwanted comments

## This PR contains valid:

- [x] tests
- [ ] permission checks (tests here too)
- [ ] translations

### CODE REVIEW

File: apps/project/tests/test_schemas.py [hunk 1]
```diff
@@ -23,13 +27,43 @@ def test_project_query(self):
...
+        LeadFactory.create(project=public_project)
+        LeadFactory.create(project=public_project)
+        LeadFactory.create(project=private_project)
+
+        # add some entry for the project
+        EntryFactory.create(project=public_project)
+        EntryFactory.create(project=public_project)
+        EntryFactory.create(project=public_project)
+        EntryFactory.create(project=public_project)
+        EntryFactory.create(project=private_project)
```
Comment (bimal125) [inline apps/project/tests/test_schemas.py]: Lets DRY

**Review by bimal125 — CHANGES_REQUESTED** [review]

File: apps/analysis_framework/schema.py [hunk 2]
```diff
@@ -101,8 +101,13 @@ class Meta:
         only_fields = (
             'id', 'title', 'description', 'is_private', 'organization',
             'created_by', 'created_at', 'modified_by', 'modified_at',
+            'preview_image'
         )
 
+    def resolve_preview_image(root, info, **kwargs):
+        if root.preview_image:
+            return info.context.request.build_absolute_uri(root.preview_image.url)
```
Comment (thenav56) [inline apps/analysis_framework/schema.py]: Let's use FileField for files.

```python
from utils.graphene.fields import FileField
...
    preview_image = graphene.Field(FileField)
```

File: apps/entry/factories.py [hunk 3]
```diff
@@ -11,6 +14,9 @@ class EntryFactory(DjangoModelFactory):
     entry_type = fuzzy.FuzzyChoice(Entry.TagType)
     excerpt = fuzzy.FuzzyText(length=100)
     image = factory.SubFactory(FileFactory)
+    project = factory.SubFactory(ProjectFactory)
+    analysis_framework = factory.SubFactory(AnalysisFrameworkFactory)
+    lead = factory.SubFactory(LeadFactory)
```
Comment (thenav56) [inline apps/entry/factories.py]: Let's pass this manually.

File: apps/project/models.py [hunk 4]
```diff
@@ -144,7 +142,26 @@ def get_for_gq(cls, user, only_member=False):
...
+                entries_count=models.functions.Coalesce(models.Subquery(
+                    Entry.objects.filter(
+                        project=models.OuterRef('pk')
+                    ).order_by().values('project').annotate(count=models.Count(
+                        'id',
+                        distinct=True
+                    )).values('count')[:1],
+                    output_field=models.IntegerField(),
+                ), 0),
+                members_count=models.Count('members', distinct=True),
```
Comment (thenav56) [inline apps/project/models.py]: Let's calculate these using data loaders instead of annotate.

File: apps/project/schema.py [hunk 5]
```diff
@@ -52,6 +56,19 @@ def resolve_allowed_permissions(root, info) -> List[str]:
...
+        if Project.objects.filter(id=root.id, members=info.context.request.user).exists():
+            return 'member'
+        elif ProjectJoinRequest.objects.filter(
+            project=root,
+            requested_by=info.context.request.user,
+            status='pending'
+        ).exists():
+            return 'pending'
+        else:
+            return 'non_member'
```
Comment (thenav56) [inline apps/project/schema.py]:
```suggestion
    @staticmethod
    def resolve_user_status(root, info):
        if root.current_user_role:
            return 'member'
        elif ProjectJoinRequest.objects.filter(
            project=root,
            requested_by=info.context.request.user,
            status='pending'
        ).exists():
            return 'pending'
        else:
            return 'non_member'
```

File: apps/project/tests/test_schemas.py [hunk 6]
```diff
@@ -23,13 +27,39 @@ def test_project_query(self):
...
+        public_project.add_member(user3)
+
+        # add some lead for the project
+        for _ in range(4):
+            LeadFactory.create(project=public_project)
+        LeadFactory.create(project=private_project)
+
+        # add some entry for the project
+        for _ in range(4):
+            EntryFactory.create(project=public_project)
```
Comment (thenav56) [inline apps/project/tests/test_schemas.py]: Use `create_batch`.

**Review by thenav56 — CHANGES_REQUESTED** [review]

File: apps/entry/factories.py [hunk 7]
```diff
@@ -3,7 +3,6 @@
 from factory.django import DjangoModelFactory
 
 from gallery.factories import FileFactory
-
```
Comment (thenav56) [inline apps/entry/factories.py]: Please add a new line. To separate relative and absolute imports.

File: apps/project/dataloaders.py [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+from utils.graphene.dataloaders import DataLoaderWithContext, WithContextMixin
+
+from project.models import Project
+
+
+class ProjectStatsLoader(DataLoaderWithContext):
+    def batch_load_fn(self, keys):
+        from lead.models import Lead
+        from entry.models import Entry
```
Comment (thenav56) [inline apps/project/dataloaders.py]: Let's move this outside.

File: apps/project/dataloaders.py [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+
+        stat_qs = Project.objects.filter(id__in=keys).annotate(
```
Comment (thenav56) [inline apps/project/dataloaders.py]: Add a `TODO: Only fetch required fields.`

File: apps/project/schema.py [hunk 10]
```diff
@@ -25,7 +31,7 @@ class Meta:
             'id', 'title', 'description', 'start_date', 'end_date',
             'regions', 'analysis_framework', 'assessment_template',
             'is_default', 'is_private', 'is_visualization_enabled', 'status',
-            'organizations', 'stats_cache',
+            'organizations', 'stats_cache', 'created_at'
```
Comment (thenav56) [inline apps/project/schema.py]: Maybe also add all this.
```
            'created_at',
            'created_by',
            'modified_at',
            'modified_by',
```

File: apps/project/schema.py [hunk 11]
```diff
@@ -34,6 +40,8 @@ class Meta:
             graphene.Enum.from_enum(PP.Permission),
         ), required=True
     )
+    project_stats = graphene.Field(ProjectStatsType)
+    user_status = graphene.Field(graphene.String)
```
Comment (thenav56) [inline apps/project/schema.py]:
```suggestion
    user_status = graphene.Field(graphene.String, required=True)
```

File: apps/project/schema.py [hunk 12]
```diff
@@ -52,6 +60,23 @@ def resolve_allowed_permissions(root, info) -> List[str]:
...
+        if root.current_user_role:
+            return 'member'
+        elif ProjectJoinRequest.objects.filter(
+            project=root,
+            requested_by=info.context.request.user,
+            status='pending'
+        ).exists():
+            return 'pending'
+        else:
+            return 'non_member'
```
Comment (thenav56) [inline apps/project/schema.py]: Let's use data loaders for ProjectJoinRequest to avoid n+1.

File: apps/project/tests/test_filters.py [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+    def test_organization_filter(self):
+        org1 = OrganizationFactory.create()
+        org2 = OrganizationFactory.create()
+        org3 = OrganizationFactory.create()
+        p1 = ProjectFactory.create()
+        p1.organizations.set([org2, org1])
+        p2 = ProjectFactory.create()
+        p2.organizations.set([org2, org3])
+        p3 = ProjectFactory.create()
+        p3.organizations.add(org1)
```
Comment (thenav56) [inline apps/project/tests/test_filters.py]: Looks like we can use create_batch here as well. Can you check?
```suggestion
        org1, org2, org3 = OrganizationFactory.create_batch(3)
        p1, p2, p3 = ProjectFactory.create_batch(3)
        p1.organizations.set([org2, org1])
        p2.organizations.set([org2, org3])
        p3.organizations.add(org1)
```

File: utils/graphene/tests.py [hunk 14]
```diff
@@ -214,6 +214,12 @@ def _include(key):
         else:
             assert excepted == real, message
 
+    def assertQuerySetEqual(self, l1, l2):
```
Comment (thenav56) [inline utils/graphene/tests.py]:
```suggestion
    def assertQuerySetIdEqual(self, l1, l2):
```

**Review by thenav56 — CHANGES_REQUESTED** [review]

File: apps/project/schema.py [hunk 15]
```diff
@@ -34,6 +40,9 @@ class Meta:
             graphene.Enum.from_enum(PP.Permission),
         ), required=True
     )
+    project_stats = graphene.Field(ProjectStatsType)
+    user_status = graphene.Field(graphene.String, required=True)
```
Comment (thenav56) [inline apps/project/schema.py]: This is not required now.

File: apps/project/tests/test_filters.py [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+        expected = [p1, p2]
+        self.assertQuerySetIdEqual(
+            expected,
+            obtained
+        )
+
+    def test_analysis_framework_filter(self):
+        af1 = AnalysisFrameworkFactory.create()
+        af2 = AnalysisFrameworkFactory.create()
+        af3 = AnalysisFrameworkFactory.create()
```
Comment (thenav56) [inline apps/project/tests/test_filters.py]: `create_batch`

File: apps/project/dataloaders.py [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+        join_status_qs = ProjectJoinRequest.objects.filter(
+            project__in=keys,
+            requested_by=self.context.request.user,
+            status='pending',
+        ).values_list('project_id', flat=True)
+        _map = {
+            project_id: 'True'
+            for project_id in join_status_qs
+        }
+        return Promise.resolve([_map.get(key, 'False') for key in keys])
```
Comment (thenav56) [inline apps/project/dataloaders.py]: Why string instead of using boolean?

**Review by thenav56 — CHANGES_REQUESTED** [review]

File: apps/project/schema.py [hunk 18]
```diff
@@ -52,6 +61,21 @@ def resolve_allowed_permissions(root, info) -> List[str]:
             root.get_current_user_role(info.context.request.user)
         )
 
+    @staticmethod
+    def resolve_project_stats(root, info, **kwargs):
+        return info.context.dl.project.project_stats.load(root.pk)
+
+    @staticmethod
```
Comment (thenav56) [inline apps/project/schema.py:111]: Remove this.

File: apps/project/schema.py [hunk 19]
```diff
@@ -34,6 +40,9 @@ class Meta:
             graphene.Enum.from_enum(PP.Permission),
         ), required=True
     )
+    project_stats = graphene.Field(ProjectStatsType)
```
Comment (thenav56) [inline apps/project/schema.py]: Remove this.

File: apps/project/schema.py [hunk 20]
```diff
@@ -15,7 +15,13 @@
...
 
 from .models import Project
-from .filter_set import ProjectFilterSet
+from .filter_set import ProjectGqlFilterSet
+
+
+class ProjectStatsType(graphene.ObjectType):
+    members_count = graphene.Field(graphene.Int)
+    sources_count = graphene.Field(graphene.Int)
+    entries_count = graphene.Field(graphene.Int)
```
Comment (thenav56) [inline apps/project/schema.py]: Remove this.

File: apps/project/dataloaders.py [hunk 21]
```diff
@@ -0,0 +1,65 @@
...
+                )).values('count')[:1],
+                output_field=models.IntegerField(),
+            ), 0),
+            members_count=models.Count('members', distinct=True),
+        )
+        _map = {
+            project_with_stat.id: project_with_stat
+            for project_with_stat in stat_qs
+        }
+        return Promise.resolve([_map.get(key) for key in keys])
```
Comment (thenav56) [inline apps/project/dataloaders.py]: Remove this.

File: apps/project/dataloaders.py [hunk 22]
```diff
@@ -0,0 +1,65 @@
...
+            for project_id in join_status_qs
+        }
+        return Promise.resolve([_map.get(key, 'False') for key in keys])
+
+
+class DataLoaders(WithContextMixin):
+
+    @cached_property
+    def project_stats(self):
+        return ProjectStatsLoader(context=self.context)
```
Comment (thenav56) [inline apps/project/dataloaders.py]: Remove this.

File: apps/project/schema.py [hunk 23]
```diff
@@ -25,7 +31,7 @@ class Meta:
             'id', 'title', 'description', 'start_date', 'end_date',
             'regions', 'analysis_framework', 'assessment_template',
             'is_default', 'is_private', 'is_visualization_enabled', 'status',
-            'organizations', 'stats_cache',
+            'organizations', 'stats_cache'
```
Comment (thenav56) [inline apps/project/schema.py]: `stats_cache` Remove this.

File: apps/project/schema.py [hunk 24]
```diff
@@ -69,7 +93,8 @@ class Meta:
             'members', 'regions', 'user_groups', 'analysis_framework',
             'category_editor', 'assessment_template', 'data',
             'is_default', 'is_private', 'is_visualization_enabled', 'status',
-            'organizations', 'stats_cache',
+            'organizations', 'stats_cache', 'created_at', 'created_by',
```
Comment (thenav56) [inline apps/project/schema.py]: Remove `stats_cache`

File: apps/project/tests/test_filters.py [hunk 25]
```diff
@@ -0,0 +1,87 @@
...
+        expected = [p2, p1]
+        self.assertQuerySetIdEqual(
+            expected,
+            obtained
+        )
+
+    def test_search_filter(self):
+        ProjectFactory.create(title='one')
+        p2 = ProjectFactory.create(title='two')
+        p3 = ProjectFactory.create(title='towo')
```
Comment (thenav56) [inline apps/project/tests/test_filters.py]: Use `create_batch`

File: apps/project/tests/test_schemas.py [hunk 26]
```diff
@@ -23,13 +28,43 @@ def test_project_query(self):
...
+                pendingMembership
               }
             }
         '''
 
         user = UserFactory.create()
+        user2 = UserFactory.create()
+        user3 = UserFactory.create()
+        request_user = UserFactory.create()
+        non_member_user = UserFactory.create()
```
Comment (thenav56) [inline apps/project/tests/test_schemas.py]: Use `create_batch`

**Review by thenav56 — CHANGES_REQUESTED** [review]

**Pull request merged**